server/services/timelineService.ts

```diff
@@ -1,28 +1,83 @@
-import type { Timeline } from 'viewModels'
-import type { TimelineResponse } from 'educationAndWorkPlanApiClient'
+import type { Prison, Timeline, TimelineEvent } from 'viewModels'
+import type { TimelineEventResponse } from 'educationAndWorkPlanApiClient'
 import EducationAndWorkPlanClient from '../data/educationAndWorkPlanClient'
 import { toTimeline } from '../data/mappers/timelineMapper'
 import logger from '../../logger'
+import PrisonService from './prisonService'
+import { HmppsAuthClient } from '../data'
+
+const SUPPORTED_TIMELINE_EVENTS = ['ACTION_PLAN_CREATED', 'INDUCTION_UPDATED']
 
 export default class TimelineService {
-  constructor(private readonly educationAndWorkPlanClient: EducationAndWorkPlanClient) {}
+  constructor(
+    private readonly hmppsAuthClient: HmppsAuthClient,
+    private readonly educationAndWorkPlanClient: EducationAndWorkPlanClient,
+    private readonly prisonService: PrisonService,
+  ) {}
 
-  async getTimeline(prisonNumber: string, token: string): Promise<Timeline> {
+  async getTimeline(prisonNumber: string, token: string, username: string): Promise<Timeline> {
     try {
+      const systemToken = await this.hmppsAuthClient.getSystemClientToken(username)
       const timelineResponse = await this.educationAndWorkPlanClient.getTimeline(prisonNumber, token)
-      const timelineResponseFilteredEvents = this.filterTimelineByActionPlanCreatedOrUpdatedEvent(timelineResponse)
-      timelineResponse.events = timelineResponseFilteredEvents
-      return toTimeline(timelineResponse)
+      timelineResponse.events = timelineResponse.events.filter(this.filterTimelineEvents)
+
+      const timeline = toTimeline(timelineResponse)
+      timeline.events = await this.addPrisonNameToPrisons(timeline.events, systemToken)
+      return timeline
     } catch (error) {
       logger.error(`Error retrieving Timeline for Prisoner [${prisonNumber}]: ${error}`)
       return { problemRetrievingData: true } as Timeline
     }
   }
 
-  filterTimelineByActionPlanCreatedOrUpdatedEvent = (timelineResponse: TimelineResponse) => {
-    return timelineResponse.events.filter(
-      (event: { eventType: string }) =>
-        event.eventType === 'ACTION_PLAN_CREATED' || event.eventType === 'INDUCTION_UPDATED',
-    )
+  private addPrisonNameToPrisons = async (
```

Comment:
OK; strap yourself in, this function is complex!
The purpose and intent of the function is to iterate through the `TimelineEvent`s and for each one lookup the prison name from it's prisonId via the `PrisonService`. It will mutate each `TimelineEvent` by setting the prison name, and will then return the mutated collection.

That's what it does .... but there is complexity! mainly because the functions in `PrisonService` are `async`, which means you need to call them with `await` and return a `Promise`.

My original code for this function was this:
```
  private addPrisonNameToPrisons = async (
    events: Array<TimelineEvent>,
    token: string,
  ): Promise<Array<TimelineEvent>> => {
    return Promise.all(
      events.map(async (event): Promise<TimelineEvent> => {
        return {
          ...event,
          prison: await this.lookupPrison(event.prison.prisonId, token),
        }
      }),
    )
  }
```

It returns a `Promise` containing all (an array) of the mapped `TimelineEvent`s. In theory and on paper it works .... however because the promise and the call to the `PrisonService` (via the `lookupPrison` function) are async the mapping of each `TmelineEvent` and the call to `PrisonService` effectively run nanoseconds apart. This in turn means we end up with a race condition where several `TimelineEvent` mappings call the `PrisonService` at "the same time".
The code in the `PrisonService` method effectively says "if its in the cache return if from the cache, else call the prison-register-api to populate the cache, then return it from the cache". Given that the "check and set cache" is not an atomic operation you can end up with a race condition, which has the effect of spamming the `prison-register-api` !!! 😬 

The only way I could get round it is to do what I've done here, and I hope the comments help 🤞